hphp/runtime/vm/jit/unique-stubs.cpp

```diff
@@ -1034,12 +1034,15 @@ TCA emitDecRefGeneric(CodeBlock& cb, DataBlock& data) {
   CGMeta meta;
 
   auto const start = vwrap(cb, data, meta, [] (Vout& v) {
+    v << vregrestrict{};
     v << stublogue{};
 
     auto const rdata = rarg(0);
     auto const rtype = rarg(1);
 
     auto const destroy = [&] (Vout& v) {
+      v << vregallow{};
```

Comment:
Ah, yes, I see.  Maybe we should give these intrinsics scoping semantics—i.e., each `vregrestrict{}` creates a nested "restrict" context that can be popped with a `vregallow{}`?  On the one hand, that seems more complicated—but on the other, it makes it less fragile, and means it's not dangerous to use in a context where we don't have full control over the entire Vunit.  Your dominator tracking can then just be incrementing and decrementing an "level" integer.